runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/internal/StateMultiplexingGroupByKey.java

```diff
@@ -0,0 +1,316 @@
+                        KV<K, V> kv = c.element();
+                        Output output = ByteString.newOutput();
+                        try {
+                          keyCoder.encode(kv.getKey(), output);
+                        } catch (IOException e) {
+                          throw new RuntimeException(e);
+                        }
+
+                        KV<ByteString, V> outputKV = KV.of(output.toByteString(), kv.getValue());
+                        if (outputKV.getKey().size() <= SMALL_KEY_BYTES_THRESHOLD) {
+                          c.output(smallKeys, outputKV);
+                        } else {
+                          c.output(largeKeys, outputKV);
+                        }
+                      }
+                    })
+                .withOutputTags(largeKeys, TupleTagList.of(smallKeys)));
+
+    PCollection<KV<K, Iterable<V>>> largeKeyBranch =
+        mapKeysToBytes
+            .get(largeKeys)
+            .setCoder(KvCoder.of(KeyedWindow.ByteStringCoder.of(), valueCoder))
+            .apply(DataflowGroupByKey.create())
+            .apply(
+                "DecodeKey",
+                MapElements.via(
+                    new SimpleFunction<KV<ByteString, Iterable<V>>, KV<K, Iterable<V>>>() {
+                      @Override
+                      public KV<K, Iterable<V>> apply(KV<ByteString, Iterable<V>> kv) {
+                        try {
+                          return KV.of(keyCoder.decode(kv.getKey().newInput()), kv.getValue());
+                        } catch (IOException e) {
+                          throw new RuntimeException(e);
+                        }
+                      }
+                    }))
+            .setCoder(outputKvCoder);
+
+    WindowFn<?, ?> windowFn = originalWindowingStrategy.getWindowFn();
+    PCollection<KV<K, Iterable<V>>> smallKeyBranch =
+        mapKeysToBytes
+            .get(smallKeys)
+            .apply(Window.into(new KeyedWindowFn<>(windowFn)))
+            .apply(
+                "MapKeys",
+                MapElements.via(
+                    new SimpleFunction<KV<ByteString, V>, KV<Integer, V>>() {
+                      @Override
+                      public KV<Integer, V> apply(KV<ByteString, V> value) {
+                        return KV.of(value.getKey().hashCode() % numShardingKeys, value.getValue());
+                      }
+                    }))
+            .apply(DataflowGroupByKey.create())
+            .apply(
+                "Restore Keys",
+                ParDo.of(
+                    new DoFn<KV<Integer, Iterable<V>>, KV<K, Iterable<V>>>() {
+                      @ProcessElement
+                      public void processElement(ProcessContext c, BoundedWindow w, PaneInfo pane) {
+                        ByteString key = ((KeyedWindow<?>) w).getKey();
+                        try {
+
+                          // is it correct to use the pane from Keyed window here?
```

Comment:
see above, I think we need to make sure we only do this for correct triggers. If that is done, I think it would be safe to use the pane.